include/triton/core/tritonbackend.h

```diff
@@ -1411,6 +1411,36 @@ TRITONBACKEND_ISPEC TRITONSERVER_Error* TRITONBACKEND_GetBackendAttribute(
     TRITONBACKEND_Backend* backend,
     TRITONBACKEND_BackendAttribute* backend_attributes);
 
+/// Query the total memory usage of the model and its associated model
+/// instances. Unless TRITONBACKEND_BackendAttributeSetEnableMemoryTracker is
+/// explicitly set to true, this function is optional, a backend is not required
+/// to implement it. This function is called when changes are made to the model
+/// and Triton wants to retrieve the latest memory usage. This function may be
+/// called multiple times within the lifetime of the model (between
+/// TRITONBACKEND_ModelInitialize and TRITONBACKEND_ModelFinalize).
+/// The backend may return error to indicate failure to get the latest memory
+/// usage.
+///
+/// \param model The model.
+/// \param usage The list of buffer attributes that the backend can record the
+/// memory usage, each entry should record the total memory usage of a given
+/// memory type and id. For example, if the model itself occupies 64 bytes on
+/// CUDA device 0 and it has two model instances that each instance occupies
+/// 128 bytes on CUDA device 1. Then 'usage' should have first two entries set,
+/// one has the buffer attributes of "type GPU, id 0, 64 bytes" and the
+/// other has "type GPU, id 1, 256 bytes". The list and the list entries are
+/// owned by Triton.
+/// \param usage_size This parameter is both input and output of the function.
+/// As input, it represents the number of entries in the list that the backend
+/// may write memory usage into. As output, it returns the number of entries
+/// that the backend has written the memory usage into. If the backend needs
+/// more entries to write the memory info, -1 should be set for 'usage_size' and
+/// error should be returned with TRITONSERVER_ERROR_INVALID_ARG.
```

Comment:
1. How does Triton core know the correct number of entries to prepare and send? via model config instance counts and such? Does this apply to backends that handle things like instance count differently like FasterTransformer or maybe KIND_MODEL in TF backend?
2. If Triton is wrong on the number of entries, should the backend instead correct Triton core on the necessary number of entries rather than just sending -1? Or is it expected via (1) that Triton core will eventually get the right number, and it can only be wrong due to some timing sensitive things like a new instance being loaded on demand while querying for the info?

Reply:
Each entry corresponds to the total memory usage of a physical device. I don't think there will be too many (~10x?). Usually if the API is designed in this way (caller owns the array), caller will expand the size and try again. We can expose a linked list structure if this is a concern.

The query is invoked after the initialize/finalize when Triton thinks the memory usage is changed. Sounds like the documentation isn't really clear on how it works?

Reply:
I haven't looked at any implementation/usage of this API yet if there is any, but just from reading these docs and your comment, my interpretation is that Triton either knows how many devices are in-use and trys with that number, or has no idea and just has some default guess, such as 10, and retries until success?:
```cpp
int32_t assumed_buffers = 10;
int32_t num_buffers = 10;

TRITONSERVER_BufferAttributes** buffer_list = new BufferList(num_buffers);
// Query memory usage with an assumed sufficient number of buffers for backend to fill
auto err = TRITONBACKEND_ModelMemoryUsage(model, buffer_list, &num_buffers);
// num_buffers == -1 or err==TRITONSERVER_INVALID_ARG, but err could be raised from other args
while (num_buffers == -1) {
  // retry with more buffers?
  num_buffers += 1
  // or num_buffers *= 2 , etc. depending on pros/cons of retries vs unnecessary allocs
}

if (assumed_buffers != num_buffers) {
  std::cout << "triton assumed " << assumed_buffers << " but backend needed: " << num_buffers << std::endl
}
```

Reply:
yes